server/neptune/workflows/internal/pr/runner.go

```diff
@@ -0,0 +1,196 @@
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/pr/receiver"
+	internalTerraform "github.com/runatlantis/atlantis/server/neptune/workflows/internal/pr/terraform"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/sideeffect"
+	"go.temporal.io/sdk/workflow"
+	"sync"
+)
+
+type Action int64
+
+const (
+	OnNewPRRevision Action = iota
+	OnShutdown
+)
+
+type RunnerState int64
+
+const (
+	Working RunnerState = iota
+	Waiting
+	Complete
+)
+
+type runnerActivities interface {
+	GithubCompareCommit(ctx context.Context, request activities.CompareCommitRequest) (activities.CompareCommitResponse, error)
+}
+
+type Runner struct {
+	RevisionSignalChannel workflow.ReceiveChannel
+	RevisionReceiver      *receiver.RevisionReceiver
+	ShutdownSignalChannel workflow.ReceiveChannel
+	ShutdownReceiver      *receiver.ShutdownReceiver
+	Activities            runnerActivities
+	Scope                 workflowMetrics.Scope
+	TFWorkflowRunner      *internalTerraform.WorkflowRunner
+
+	// mutable state
+	state                 RunnerState
+	lastProcessedRevision string
+}
+
+func newRunner(ctx workflow.Context, scope workflowMetrics.Scope, tfWorkflowRunner *internalTerraform.WorkflowRunner, a *prActivities) *Runner {
+	revisionReceiver := receiver.NewRevisionReceiver(ctx, scope)
+	shutdownReceiver := receiver.NewShutdownReceiver(ctx, scope)
+	return &Runner{
+		RevisionSignalChannel: workflow.GetSignalChannel(ctx, receiver.TerraformRevisionSignalID),
+		RevisionReceiver:      &revisionReceiver,
+		ShutdownSignalChannel: workflow.GetSignalChannel(ctx, receiver.ShutdownSignalID),
+		ShutdownReceiver:      &shutdownReceiver,
+		Activities:            a,
+		Scope:                 scope,
+		TFWorkflowRunner:      tfWorkflowRunner,
+	}
+}
+
+// Run handles managing the workflow's context lifecycles as new signals/poll events are received and
+// change the current PRAction status
+func (r *Runner) Run(ctx workflow.Context) error {
+	var action Action
+	var prRevision receiver.Revision
+
+	//TODO: add approve signal, timeouts, poll variations of signals
+	s := workflow.NewSelector(ctx)
+	s.AddReceive(r.RevisionSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		prRevision = r.RevisionReceiver.Receive(c, more)
+		action = OnNewPRRevision
+	})
+
+	s.AddReceive(r.ShutdownSignalChannel, func(c workflow.ReceiveChannel, more bool) {
+		r.ShutdownReceiver.Receive(c, more)
+		action = OnShutdown
+		r.state = Complete
+	})
+
+	inProgressCtx, inProgressCancel := workflow.WithCancel(ctx)
+	for {
+		s.Select(ctx)
+		switch action {
+		case OnNewPRRevision:
+			shouldProcess, err := r.shouldProcessRevision(ctx, prRevision)
+			if err != nil {
+				return err
+			}
+			if !shouldProcess {
+				continue
+			}
+			// cancel in progress workflow (if it exists) and start up a new one
+			inProgressCancel()
+			inProgressCtx, inProgressCancel = workflow.WithCancel(ctx)
```

Comment:
if we have to start naming our context something other than ctx that's a sign we need to move this code to another function.